File: examples/python/location_search.py
#!/usr/bin/env python3
"""Demonstrate `search_locations` only using Normatia's search endpoint."""

from __future__ import annotations

import asyncio
import os
import sys
from typing import Any

import httpx

DEFAULT_API_URL = "https://api.normatia.com"


def get_client_config() -> tuple[str, dict[str, str]]:
    api_key = os.getenv("NORMATIA_API_KEY")
    if not api_key:
        raise RuntimeError(
            "Missing NORMATIA_API_KEY environment variable. "
            "Set it to your bearer key, for example: sk-normatia-xxxxx"
        )

    base_url = os.getenv("NORMATIA_API_URL", DEFAULT_API_URL).rstrip("/")
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Accept": "application/json",
    }
    return base_url, headers


def print_location_results(results: list[dict[str, Any]]) -> None:
    print("Location Search Results")
    print("=======================")
    if not results:
        print("No locations found.")
        return

    for index, item in enumerate(results, start=1):
        ancestors = item.get("ancestors") if isinstance(item, dict) else []
        if not isinstance(ancestors, list):
            ancestors = []

        ancestor_names = [
            ancestor.get("name", "")
            for ancestor in ancestors
            if isinstance(ancestor, dict)
        ]

        print(f"{index}. geo_id: {item.get('geo_id')}")
        print(f"   name: {item.get('name')}")
        print(f"   level: {item.get('level')}")
        print(f"   ancestors: {', '.join(filter(None, ancestor_names)) or 'n/a'}")


async def main() -> None:
    base_url, headers = get_client_config()

    try:
        async with httpx.AsyncClient(base_url=base_url, headers=headers, timeout=20.0) as client:
            search_response = await client.get("/api/v1/location/search", params={"q": "Sevilla"})
            search_response.raise_for_status()
            search_payload = search_response.json()

            results: list[dict[str, Any]] = []
            if isinstance(search_payload, dict) and isinstance(search_payload.get("results"), list):
                results = [item for item in search_payload["results"] if isinstance(item, dict)]

            print_location_results(results)

    except httpx.HTTPStatusError as exc:
        print("\nRequest failed with a non-success HTTP status.")
        print(f"Status: {exc.response.status_code}")
        print(f"URL: {exc.request.url}")
        print(f"Body: {exc.response.text}")
        sys.exit(1)
    except httpx.RequestError as exc:
        print("\nNetwork error while calling Normatia API.")
        print(f"Error: {exc}")
        sys.exit(1)
    except RuntimeError as exc:
        print(f"\nConfiguration or response error: {exc}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/location-search.ts
/**
 * Demonstrates Normatia Location API usage with raw fetch.
 */

const DEFAULT_API_URL = "https://api.normatia.com";

type ClientConfig = {
  apiKey: string;
  baseUrl: string;
};

function getClientConfig(): ClientConfig {
  const apiKey = process.env.NORMATIA_API_KEY;
  if (!apiKey) {
    console.error("Error: NORMATIA_API_KEY is not set.");
    console.error("Set it first: export NORMATIA_API_KEY='sk-normatia-xxxxx'");
    process.exit(1);
  }

  return {
    apiKey,
    baseUrl: (process.env.NORMATIA_API_URL ?? DEFAULT_API_URL).replace(/\/+$/, ""),
  };
}

function buildRequest(
  config: ClientConfig,
  path: string,
  init: RequestInit = {},
): { url: string; init: RequestInit } {
  const headers = new Headers(init.headers);
  headers.set("Authorization", `Bearer ${config.apiKey}`);
  headers.set("Accept", "application/json");

  return {
    url: `${config.baseUrl}${path}`,
    init: { ...init, headers },
  };
}

function withQuery(path: string, params: Record<string, string>): string {
  const query = new URLSearchParams(params);
  return `${path}?${query.toString()}`;
}

function formatOutput(value: unknown): string {
  if (typeof value === "string") {
    return value;
  }
  return JSON.stringify(value, null, 2);
}

async function fetchJson(config: ClientConfig, path: string): Promise<unknown> {
  const request = buildRequest(config, path, { method: "GET" });

  let response: Response;
  try {
    response = await fetch(request.url, request.init);
  } catch (error) {
    throw new Error(`Network error while calling ${request.url}: ${String(error)}`);
  }

  const raw = await response.text();
  let payload: unknown = null;
  if (raw) {
    try {
      payload = JSON.parse(raw);
    } catch {
      payload = raw;
    }
  }

  if (!response.ok) {
    throw new Error(
      `Request failed (${response.status} ${response.statusText}) for ${request.url}\n${formatOutput(payload)}`,
    );
  }

  return payload;
}

async function runCall(config: ClientConfig, title: string, path: string): Promise<void> {
  console.log(`\n== ${title} ==`);
  const payload = await fetchJson(config, path);
  console.log(formatOutput(payload));
}

async function main(): Promise<void> {
  const config = getClientConfig();

  await runCall(
    config,
    "1) Basic search: Sevilla",
    withQuery("/api/v1/location/search", { q: "Sevilla" }),
  );

  await runCall(
    config,
    "2) Filter by level: municipality",
    withQuery("/api/v1/location/search", {
      q: "Sevilla",
      level: "municipality",
    }),
  );

  await runCall(
    config,
    "3) Filter by ancestor_id: municipalities inside Sevilla province (ES-41)",
    withQuery("/api/v1/location/search", {
      q: "Mairena del Aljarafe",
      ancestor_id: "ES-41",
      level: "municipality",
    }),
  );

  await runCall(config, "4) Location detail: ES-41091 (Sevilla municipality)", "/api/v1/location/ES-41091");
}

main().catch((error) => {
  console.error("\nLocation example failed.");
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
});
